kernelCTF: Add CVE-2024-41010_lts (#126)
google/security-research: pocs/linux/kernelctf/CVE-2024-41010_lts/docs/novel_techniques.md

			rebalance = NULL;
		} else {
			rebalance = rb_is_black(successor) ? parent : NULL;
		}
		successor->__rb_parent_color = pc;
		tmp = successor;
	}

	augment->propagate(tmp, NULL);
	return rebalance;
}
```

The `rb_set_parent_color()` doesn't have any validation of linked Rbtree corruption, so unsafe unlink is possible. It makes write any accessable kernel address to any kernel address. Furthermore, the kernel doesn't panic eventhough the corruption of Rbtree node. It makes various type of exploitation potentially. 
```cpp
static inline void
__rb_change_child(struct rb_node *old, struct rb_node *new,
		  struct rb_node *parent, struct rb_root *root)
{
	if (parent) {
		if (parent->rb_left == old)
			WRITE_ONCE(parent-(>rb_left, new);
		else)
			WRITE_ONCE(parent->rb_right, new);
	} else
		WRITE_ONCE(root->rb_node, new);
}
```

The exp183 use this premitive in two way.
1. The corruption of Rbtree node data doesn't trigger kernel panic, so freely free corrupted `struct simple_xattr` without any consideration.
2. Overwrite 1 byte of `node->rb_node.rb_right` to `0x00`(make fake_chunk) and use fake_chunk freely for triggering double_free and UAF.

> ## Oneshot in Kernel: WakeROP

Reviewer: This gadget exists since [kernel 2.4](https://elixir.bootlin.com/linux/2.4.22/source/arch/x86_64/kernel/acpi_wakeup.S#L183), x86_64.

> ## Oneshot in Kernel: WakeROP

Reviewer 2: Done

The RIP Pivoting not that powerful because we usually need ROP to exploit. Also, we need to use much time for finding proper gadget and it is changed everytime. We usually solve this problem only depended on the register and a few proper gadget. 

> However, there are exist powerful gadget in `wakeup_long64()`.

Reviewer 2: ```suggestion
However, there is a powerful gadget in `wakeup_long64()`.
```

```asm
SYM_FUNC_START(wakeup_long64)
	movq	saved_magic, %rax
	movq	$0x123456789abcdef0, %rdx
	cmpq	%rdx, %rax
	je	2f

	/* stop here on a saved_magic mismatch */
	movq $0xbad6d61676963, %rcx
1:
	jmp 1b
2:
	movw	$__KERNEL_DS, %ax
	movw	%ax, %ss	
	movw	%ax, %ds
	movw	%ax, %es
	movw	%ax, %fs
	movw	%ax, %gs
	movq	saved_rsp, %rsp

	movq	saved_rbx, %rbx
	movq	saved_rdi, %rdi
	movq	saved_rsi, %rsi
	movq	saved_rbp, %rbp

	movq	saved_rip, %rax
	ANNOTATE_RETPOLINE_SAFE
	jmp	*%rax
SYM_FUNC_END(wakeup_long64)
```

> This code copy general purpose register from some variable, even `rip` and `rsp`. When you see this code in vmlinux, copy all register from code's `bss` and jump to `rax` address.

Reviewer 2: ```suggestion
This code copies general purpose registers from some variable, even `rip` and `rsp`. When you see this code in vmlinux, copy all registers from code's `bss` and jump to `rax` address.
```

Therefore, only write some address to `saved_*` section makes full control of system.
```
   0xffffffff8112c291 <wakeup_long64+49>:       mov    rsp,QWORD PTR ds:0xffffffff83c51a68
   0xffffffff8112c299 <wakeup_long64+57>:       mov    rbx,QWORD PTR ds:0xffffffff83c51a58
   0xffffffff8112c2a1 <wakeup_long64+65>:       mov    rdi,QWORD PTR ds:0xffffffff83c51a50
   0xffffffff8112c2a9 <wakeup_long64+73>:       mov    rsi,QWORD PTR ds:0xffffffff83c51a48
   0xffffffff8112c2b1 <wakeup_long64+81>:       mov    rbp,QWORD PTR ds:0xffffffff83c51a40
   0xffffffff8112c2b9 <wakeup_long64+89>:       mov    rax,QWORD PTR ds:0xffffffff83c51a60
   0xffffffff8112c2c1 <wakeup_long64+97>:       jmp    rax
```

> There are so many ways to apply this technique to exploit. exp184 also use this technique for exploiting.

Reviewer 2: Could you please add part of diagram from slide 118 (https://www.hexacon.fr/slides/Cho_Lee-Utilizing_Cross-CPU_Allocation_to_Exploit_Preempt-Disabled_Linux_Kernel.pdf) which illustrates the fake_stack that should be constructed for WakeROP?

This technique is helpful for below situation.

0. Generally,
	- it can be used when there is insufficient ROP space.
	- it can be used when there is no proper gadget. (by calling `set_memory_x(saved_*)` and make arbitrary gadget)
1. UAF & Cross Cache Avaliable
	- use [PIPEShot](#oneshot-in-kernel-wakerop): kernelctf exp183
2. Only know kbase (by leak or side channel) & RIP pivoting
	- use CEA (`write_cpu_entry_area`) + WakeROP: kernelctf exp184

## PIPEShot
Only [OneShot in Kernel: WakeROP](#oneshot-in-kernel-wakerop) is strong enough to exploit, but not general. We want to introduce PIPEShot, which is generally utilized for UAF exploit

1. cross cache is available
	- generally used for all SLAB UAF
2. cross cache is not availablex

> 	- generally used for `kmalloc-cg-<64, 192, 512, and over>` SLAB UAF

Reviewer 2: Could you provide more details on why, when the cross-cache is not available, we could still use this technique? How have you determined that those specific caches would work?

Reviewer 2: Could you elaborate why those specific caches would work? Why not 256 for example?

The main flow of PIPEShot is below.

0. Let we know `kbase` and `kheapbase` address
	- `kbase` is easily leaked by UAF or side channel
	- `kheapbase` is easily leaked when UAF is available
1. Create pipe
	- let `pipeshot_pipe`
```
 ----------- pipeshot_pipe -----------
|                                     |
|                                     |
 -------------------------------------
```
2. Trigger UAF
3. Alloc any object which user can R/W (e.g. `struct simple_xattr` or `struct msg_msgseg`)

> 	- let `pipeshot_data`, then `pipeshot_pipe` and `pipeshot_data` are duplicated

Reviewer 2: Could you cover this step in more detail? What do you mean by duplicated? What is the example memory layout in this case?

```
 ----------- pipeshot_pipe ----------- ------------ pipeshot_data ----------
|                                     |                                     |
|                                     |                                     |
|                                     |                                     |
|                                     |                                     |
 ------------------------------------- -------------------------------------
```
4. Write to pipe

> 	- If user write some data to pipe, save `VMEMMAP_BASE + alpha` addr to `pipeshot_pipe->buffer`

Reviewer 2: Could you add more information on what data user should write? How / from which structures / fields user could collect `VMEMMAP_BASE + alpha` addr?
